rustc_codegen_spirv/src/link.rs

```diff
@@ -124,97 +128,86 @@ fn link_exe(
+
+    let result = optimizer.optimize(
+        &spv_binary,
+        &mut |msg: error::Message| {
+            use error::MessageLevel as Level;
+
+            // TODO: Adds spans here? Not sure how useful with binary, but maybe?
+
+            let mut err = match msg.level {
+                Level::Fatal | Level::InternalError => sess.struct_fatal(&msg.message),
+                Level::Error => sess.struct_err(&msg.message),
+                Level::Warning => sess.struct_warn(&msg.message),
+                Level::Info | Level::Debug => sess.struct_note_without_error(&msg.message),
+            };
+
             err.note(&format!("module {:?}", filename));
-            err
+            err.emit();
         },
+        // We currently run the validator separately after optimization or even
+        // if we don't run optimization, the default options don't run the validator
+        None,
     );
-}
 
-/// Runs a given SPIR-V `tool`, configured with `builder`, erroring if not found
-/// or returns a non-zero exit code. All errors will be emitted using the
-/// diagnostics builder provided by `diagnostics.`
-fn output_spirv_tool<'a, F, D>(
-    sess: &Session,
-    tool: &str,
-    builder: F,
-    diagnostics: D,
-) -> Option<std::process::Output>
-where
-    F: FnOnce(&mut Command) -> &mut Command,
-    D: FnOnce(std::process::ExitStatus) -> DiagnosticBuilder<'a>,
-{
-    let mut cmd = Command::new(tool);
-    (builder)(&mut cmd);
-
-    let output = match cmd.output() {
-        Ok(output) => output,
+    match result {
+        Ok(binary) => Vec::from(binary.as_ref()),
         Err(_) => {
-            let mut err =
-                sess.struct_err(&format!("Couldn't find `{}` SPIR-V tool in PATH.", tool));
-            err.note(
-                "Please ensure that you have `spirv-tools` installed on \
-                      your system and available in your PATH.",
-            );
-            err.emit();
-            return None;
+            let mut err = sess.struct_warn("spirv-opt failed, leaving as unoptimized");
+            err.note(&format!("module {:?}", filename));
+            spv_binary
         }
-    };
+    }
+}
 
-    if output.status.success() {
-        Some(output)
-    } else {
-        let mut diagnostics = (diagnostics)(output.status);
+fn do_spirv_val(sess: &Session, spv_binary: &[u32], filename: &Path) {
+    use spirv_tools::val::{self, Validator};
 
-        if !output.stdout.is_empty() {
-            diagnostics.note(&String::from_utf8(output.stdout).unwrap());
-        }
-        if !output.stderr.is_empty() {
-            diagnostics.note(&String::from_utf8(output.stderr).unwrap());
-        }
+    let validator = val::create(None);
 
-        diagnostics.emit();
-        None
+    if validator.validate(spv_binary, None).is_err() {
+        let mut err = sess.struct_err("error occurred during validation");
+        err.note(&format!("module {:?}", filename));
+        err.emit();
```

Comment:
just checking, does this emit the error message? maybe the tool already prints it? (maybe add a comment if it does print it, since this is a little confusing if so)